bundles/org.openhab.binding.freeathomesystem/src/main/java/org/openhab/binding/freeathomesystem/internal/FreeAtHomeSystemHandlerFactory.java

```diff
@@ -0,0 +1,131 @@
+public class FreeAtHomeSystemHandlerFactory extends BaseThingHandlerFactory {
+
+    private final Logger logger = LoggerFactory.getLogger(FreeAtHomeSystemHandlerFactory.class);
+
+    private @NonNullByDefault({}) HttpClient httpClient;
+    private final FreeAtHomeThingTypeProvider thingTypeProvider;
+    private final FreeAtHomeChannelTypeProvider channelTypeProvider;
+    private final FreeAtHomeChannelGroupTypeProvider channelGroupsTypeProvider;
+
+    private void generteThingTypes() {
+        String label = "free-at-home-device";
+        String description = String.format("Generic free@home device");
+
+        List<String> supportedBridgeTypeUids = new ArrayList<>();
+        supportedBridgeTypeUids.add(BRIDGE_TYPE_UID.toString());
+        ThingTypeUID thingTypeUID = new ThingTypeUID(BINDING_ID, FREEATHOMEDEVICE_TYPE_ID);
+
+        Map<String, String> properties = new HashMap<>();
+        properties.put(Thing.PROPERTY_VENDOR, "Busch&Jaeger - ABB");
+        properties.put(Thing.PROPERTY_MODEL_ID, "free@home Device");
+
+        URI configDescriptionURI;
+        try {
+            configDescriptionURI = new URI("thing-type:freeathomesystem:free-at-home-device");
+
+            ThingType thingType = ThingTypeBuilder.instance(thingTypeUID, label)
+                    .withSupportedBridgeTypeUIDs(supportedBridgeTypeUids).withDescription(description)
+                    .withProperties(properties).withConfigDescriptionURI(configDescriptionURI).build();
+
+            this.thingTypeProvider.addThingType(thingType);
+        } catch (URISyntaxException e) {
+            logger.debug("Exception during creating config description URI");
+        }
+
+        return;
+    }
+
+    @Activate
+    public FreeAtHomeSystemHandlerFactory(@Reference FreeAtHomeThingTypeProvider thingTypeProvider,
+            @Reference FreeAtHomeChannelTypeProvider channelTypeProvider,
+            @Reference FreeAtHomeChannelGroupTypeProvider channelGroupsTypeProvider) {
+        this.thingTypeProvider = thingTypeProvider;
+        this.channelTypeProvider = channelTypeProvider;
+        this.channelGroupsTypeProvider = channelGroupsTypeProvider;
+
+        generteThingTypes();
+    }
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (BRIDGE_TYPE_UID.equals(thingTypeUID)) {
+            logger.debug("Create SysAP bridge");
+            return new FreeAtHomeBridgeHandler((Bridge) thing, httpClient);
+        } else if (FREEATHOMEDEVICE_TYPE_UID.equals(thingTypeUID)) {
+            logger.debug("Create free@home device");
+            return new FreeAtHomeDeviceHandler(thing, channelTypeProvider, channelGroupsTypeProvider);
+        }
+
+        return null;
+    }
+
+    @Reference
```

Comment:
Why this approach rather than constructor injection? See for example https://github.com/openhab/openhab-addons/pull/15003/files#diff-98646f92898bea7351ac2044172ae75db57f72c3d53757a1763be1e4a52fd6c2R54-R66

Additionally, any reason not to manage life cycle here, i.e. start and stop the client?

Reply:
@jlaur 
Thank the hint regarding the ThingHandlerService. However it is not working for me (yet).
If I make the same implementation as the Indego binding the SCAN button disappears and the method setThingHandler is never called.
Could you give a further hint what could be missing?

Reply:
Just to be clear, this comment was about getting `httpClientFactory` through constructor injection in this class.

I assume your question is about discovery? I would need some more context in order to provide feedback.

Reply:
Hi @jlaur,

The httpClient class injection via constructor is already implemented. 

You notified correctly I mentioned the discovery service which you mentioned in the [comment](https://github.com/openhab/openhab-addons/pull/13703/files/7993717ae63fc2fe186fe2c8161d1edb6a273255#r1275032628) below

However this code snippet as seen in the mentioned link having some interaction in Bosch Indego binding. Therefore I added here my question.

If I understand correctly in the Bosch Indego binding, to use the ThingHandlerService in the discovery service class I need to add some activation in the Handler Factors class (at least it is implemented in the Bosch binding)
However I did not understand the mechanism and the result is now the the discovery is not working.

Sorry for the confusion, maybe I added my question to the wrong place. however the two classes are looking for me as connected for the ThingHandlerService.

If you have some idea how I am able to use the ThingHandlerService, please let me know.

Thank you in advance 


If I implement

Reply:
I can't see anything wrong in the discovery class at first glance that would explain it not working at all. Perhaps it would be worth to post in community forum for getting help from a broader audience, since I'm only a single reviewer here. At first I would suggest to figure out which methods are called and get more details, since "not working" is not very concrete. 🙂

Reply:
Maybe this could be the missing part (in the bridge handler):
https://github.com/openhab/openhab-addons/blob/297640aa774556cfbf70044321ff90bd3b21e8d1/bundles/org.openhab.binding.boschindego/src/main/java/org/openhab/binding/boschindego/internal/handler/BoschAccountHandler.java#L151-L154

Reply:
> Additionally, any reason not to manage life cycle here, i.e. start and stop the client?

I think this part was missed. The client is started, configured and stopped inside the bridge handler in `openHttpConnection` and `closeHttpConnection`. Those methods look a bit over-complicated, is there any reason not to manage the life cycle directly here so that the handler doesn't have to deal with this?

It's a quite common used pattern, and you can also let the binding fail early in case you are not able to start the client. Example:
https://github.com/openhab/openhab-addons/blob/e62bcf5db4f3e3126f60ea467d31cbdd9755ee49/bundles/org.openhab.binding.http/src/main/java/org/openhab/binding/http/internal/HttpHandlerFactory.java#L64-L71

Reply:
I still don't see why you don't configure and start the `HttpClient` within the `FreeAtHomeSystemHandlerFactory` constructor? To be specific, I mean only these lines (modified error handling):
```java
        httpClient = httpClientFactory.createHttpClient(FreeAtHomeSystemBindingConstants.BINDING_ID);
        httpClient.setFollowRedirects(false);
        httpClient.setMaxConnectionsPerDestination(1);
        httpClient.setMaxRequestsQueuedPerDestination(50);
        httpClient.setIdleTimeout(-1);
        httpClient.setConnectTimeout(5000);

        try {
            httpClient.start();
        } catch (Exception e) {
            throw new IllegalStateException("Could not create HttpClient", e); 
        }
```

and:
```java
    @Deactivate
    public void deactivate() {
        try {
            httpClient.stop();
        } catch (Exception e) {
            // catching exception is necessary due to the signature of HttpClient.stop()
            logger.warn("Failed to stop HttpClient: {}", e.getMessage());
        }
    }
```

This way the bridge handler wouldn't have to deal with trying to handle this error. It's also unlikely it will be able to recover from this. The binding won't work without being able to start the HttpClient anyway.

Reply:
My design decision at implementation was to move everything belongs to the bridge, HTTP and Websocket communication to the Bridge Handler.
What you propose means that the HTTP Client configuration, start/stop is in the Factory but the GET/SET Methods are in the Bridge handler.

But let's how it could look like if I move to the start/stop to the Factory

Reply:
The proposal is only to configure and start to client in the factory, but not do any kind of communication - that should be kept in the bridge handler.

Reply:
OK.. I move it tomorrow to the Factory

Reply:
**Fixed**

I took over the proposal

Reply:
It looks good! 👍